### ISSUE TITLE
Match algorithm is starving some bots

### ISSUE THREAD

**Issue by zaphus** [aichallenge/aichallenge#179]
I've been watching the top few bots, and have noticed that some are not getting a game in a long long time.

Zaphus (mine) has gone 10 hours without a game
George (someone elses) has gone 15 hours without a game

others have had many many games in that time.  Same issue happened on PW and it drove contestants batty

Side note:  In the 10 hours since my bot played, my Trueskill ranking dropped from 54 to 49 - even though I was ranked #1 at the start of that time.  That won't make much sense to the contestants

**Comment by McLeopold** [comment]
I was actually slightly concerned about this.  Right now there is some no repeat logic in place where you aren't allowed to play the same bot if you already had a match with then in the last 5 games.  This may be too restrictive with the current low number of active bots.  I suspect the matches have poor match quality right now (according to trueskill.)

As far as the bot rotation goes, the seed player logic should be working just fine.  The additional players should favor older submissions, so newer submissions wouldn't get as many games.  Would you be willing to analyze this if you had an sql dump?

**Comment by zaphus** [comment]
Yeah, I'll take a look at the dump if you get it to me.
I don't believe the seeding is working, since (if I understand it) it should play every bot once as the seed before doing any one bot gets seeded twice.  I can find gaps of over 200 games for some bots, when there are only 60 bots in the ranking (and 200 games means at least 800 bots were chosen to play in that time)

**Comment by McLeopold** [comment]
The seed values are indeed incorrect.  A player is seeded several times in a row.

**Comment by McLeopold** [comment]
You'll want to look at the 02_generate_matchup sql.  It's fairly complicated, but choosing the seed value is at the top, so you can ignore everything past step 1.  The game table contains completed games and the matchup table contains games that are in progress or will be sent to the worker.

The matchup table does not get cleaned up properly if a matchup fails.  There is an error code I believe.  We want the matchup data to stick around, but I think it is then affecting the generate matchup code.  The other thought is that generate matchup is ignoring games in progress.

**Comment by Janzert** [comment]
Another example, June 18 from 23:32 to 23:42 FlagCapper received 13 games. http://aichallengebeta.hypertriangle.com/profile_games.php?user=7&page=1

**Comment by zaphus** [comment]
I'm looking into this, but my SQL knowledge is low.   Will be talking to an SQL expert today to see if he agrees with what I suspect the problem is, and if he can show me the correct SQL commands to do what will be a better method

**Comment by McLeopold** [comment]
Not sure why this was closed.  I haven't seen a fix yet.

**Comment by zaphus** [comment]
I really hate "Comment and close" because I keep clicking on it thinking it is going to add my comment and close the window, not close the issue :-)   I'll get there eventually!
Meanwhile...  I'll see if I can figure something out today - but my SQL knowledge is too small to even get the data I want out of the DB right now, trying to learn fast

**Comment by TRavary** [comment]
An idea of the reason for the difference of rate between players.
In step 3 of "2_generate_matchup.sql", the next player in matchup is chosen to have the best match quality among the 100 closest players in terms of mu.
More is sigma, less is the quality. So players with high sigma (relatively with close players) will be less chosen.
Look at high rate players (margusja 93th, hbhasker2 85th, Parsley 69th...) they all have high sigma.

**Comment by TRavary** [comment]
I made a test with the help of Bubumiki which play twice more game than me (3h to play 10 games vs 8h for me). We uploaded the same bot at the same time. The 2 bots are climbing the rankings with similar trueskill. but Bubumiki always play twice more than me.

http://aichallengebeta.hypertriangle.com/profile.php?user=118        (version 7)
http://aichallengebeta.hypertriangle.com/profile.php?user=355        (version 45)

Another criterion than trueskill cause the difference of numbers games played.

(this comment is maybe connected with issue #215 :"top player dont play agains each other")
